src/coreclr/jit/lsraxarch.cpp

```diff
@@ -2481,8 +2481,15 @@ int LinearScan::BuildHWIntrinsic(GenTreeHWIntrinsic* intrinsicTree)
                         // When op2 is not contained or if we are producing a scalar value
                         // we need to mark it as delay free because the operand and target
                         // exist in the same register set.
-
-                        srcCount += BuildDelayFreeUses(op2);
+                        // Unless, op1 and op2 are same, in which case we can overwrite op2.
+                        if (GenTree::NodesAreEquivalentLeaves(op1, op2))
```

Comment:
Are there other RMW locations setting `BuildDelayFreeUses` that could/should also be updated? Is this already handled on the scalar operation code path, for example?

Reply:
I'd imagine, for example, the check a few lines down dealing with making `BuildDelayFreeUses(op3)` would also benefit from this check: https://github.com/dotnet/runtime/pull/53964/files#diff-9626112837daf480c93d401a587012b3e398dd90a953c870797d472cff36839dR2510

Maybe the right fix is to call this overload of `BuildDelayFreeUses`: https://github.com/dotnet/runtime/blob/c2c92c4cf3e650d192cc39da89ba130aef698bfa/src/coreclr/jit/lsrabuild.cpp#L3059
The signature is: `int LinearScan::BuildDelayFreeUses(GenTree* node, GenTree* rmwNode, regMaskTP candidates)` and it looks to already do some checks on `rmwNode` vs the `node` being set as `delayFree` and might be a natural place to add this `NodesAreEquivalentLeaves` check, if its generally applicable?

Reply:
That's right, at most places, we pass `rmwNode` as nullptr and I am not 100% sure the impact of refactoring it such that we can fit in `NodesAreEquivalent()` method inside it.  For this bug, I would keep it where it is currently.

> a few lines down dealing with making BuildDelayFreeUses(op3) would also benefit from this 

Can you give an example of what is the semantics of rmw for this and which 2 nodes I should be checking for equivalence?

Reply:
> I am not super familiar with the RMW semantics.

`RMW` (`Read-Modify-Write`) just means one of the sources is also the destination and so is "destructive" (the value isn't preserved).

This means that if the value needs to be preserved (generally because `src1` is not "last use") that you need an additional `mov` to copy the value to the destination:
```asm
; Operation is logically `dst = src1 + src2`
mov dst, src1
add dst, src2
```

Because of this, it also means that `dst != src2` since, if it was, the move would overwrite `src2` and change the logical operation. `BuildDelayFreeUses` exists to cover this scenario by ensuring that `dst != src2`:
```asm
; If `dst == src2 && src1 != src2`, operation becomes `dst = src1 + src1` and so is problematic
mov dst, src1
add dst, src2
```

All of this works great when `src1` is last use and when `src1 != src2` because this means the register allocator can make `dst == src1` and we can elide the copy generating just:
```asm
; Operation is logically `src1 += src2`
add src1, src2
```

However, in certain cases such as when `src1 == src2`, setting `src2` to be "delay free" is problematic because this forces the register allocator to configure it as `dst != src2`, but since `src1 == src2` this also means the register allocator configures `dst != src1`, thus forcing us to generate a move. In practice however, it is safe for this exact scenario to not be delay free because we would never overwrite either input.

The PR you are providing here "fixes" the issue by avoiding us setting "delay free" when `rmwOp == delayFreeOp` (this is safe since it means we won't ever overwrite `delayFreeOp`). This "should be" applicable to all RMW setups whether its specifically this SIMD example or even other examples like integer additions and so I'd think that we want to make the support for this "broader" so we can always generate the more efficient code.

As far as I can tell, the fix you have here should work for any `BuildDelayFreeUses` call and so *I think* we could just extend most calls to `BuildDelayFreeUses` to pass in `rmwOp` when it exists and for it to do the `if (GenTree::NodesAreEquivalentLeaves(delayFreeOp, rmwOp)) { /* dont set delay free */ } else { /* set delay free */ }`
* For the `2-operand` RMW instructions (like `add`, `sub`, `mul`, `div`, etc) this is what you are already doing for this specific SIMD case.
* It should likewise extend to `3-operand` (or more) RMW instructions (like `fma` where you have `dst = (src1 * src2) + src3`). In the three operand case, eliding the move requires: `(dst == src1) && (dst != src2) && (dst != src3)`. However the latter two restrictions (which are achieved by setting "delay free") can be relaxed when `src1 == src2` or when `src1 == src3`, respectively (since when `rmwOp == delayFreeOp`, we can't overwrite the `delayFreeOp`).

Reply:
Noting that I'm not super familiar with the register allocator logic myself, just stating logically how I'd expect this to be handled, and so it isn't clear to me if this is already "being handled" by the https://github.com/dotnet/runtime/blob/c2c92c4cf3e650d192cc39da89ba130aef698bfa/src/coreclr/jit/lsrabuild.cpp#L3100 check.

If it is, it might be simple enough to just ensure we pass in `rmwOp` everywhere and the right things will happen.

Reply:
Thanks for the wonderful explanation @tannergooding . I think I heard from Carol about `delayRegFree`, but your concrete examples helped me understand it more clearly. Regarding, `RMW` (last I heard was when working on Arm64), I thought that only certain instructions fall under that category (and that's why was asking which one are those), but looks like it is applicable to even GP register cases like `add`. I will investigate and try to come up with "broader" solution.

Reply:
Also, from what I understand, `rmwOp` is essentially the one that is acting as source as well as destination operand, while other operand (e.g. `node` parameter in `BuildDelayFreeUses()`) refers to other operand of that operation. I need to look deeper, but what determines what should be `node` parameter to `BuildDelayFreeUses()`? Assuming `op1` is always considered as potential source and destination for 2/3/4 operand cases (and hence it is `rmwNode`), but for 3+ operand cases, should `node` be `op2` or `op3` or it depends on the `GenTree`?

Reply:
> but looks like it is applicable to even GP register cases like add

Right. On ARM64 most instructions are not RMW, that is they can separately encode `dst`, `src1`, and `src2`. Only a handful of instructions end up being `RMW` and many of them are more advanced SIMD instructions so its a rarity to deal with.
x86/x64 on the other hand has most instructions being RMW and outside some newer instructions or using the VEX encoding, its something that codegen frequently has to deal with.

> Also, from what I understand, rmwOp is essentially the one that is acting as source as well as destination operand, while other operand (e.g. node parameter in BuildDelayFreeUses()) refers to other operand of that operation

Right. `rmwOp` is the operand that is a source but also the destination.

> Assuming op1 is always considered as potential source and destination for 2/3/4 operand cases (and hence it is rmwNode), but for 3+ operand cases, should node be op2 or op3 or it depends on the GenTree?

We have to build uses for all operands and so typically we:
* Set the `rmwOp` (which is typically, but not always, `op1`) to be `tgtPrefUse`: https://github.com/dotnet/runtime/blob/c2c92c4cf3e650d192cc39da89ba130aef698bfa/src/coreclr/jit/lsraxarch.cpp#L2455
* For each other operand, we call `BuildDelayFreeUses`: https://github.com/dotnet/runtime/blob/c2c92c4cf3e650d192cc39da89ba130aef698bfa/src/coreclr/jit/lsraxarch.cpp#L2491 and https://github.com/dotnet/runtime/blob/c2c92c4cf3e650d192cc39da89ba130aef698bfa/src/coreclr/jit/lsraxarch.cpp#L2510

There are many examples of this throughput lsraxarch.cpp

Reply:
One of the 4 operand examples is here: https://github.com/dotnet/runtime/blob/c2c92c4cf3e650d192cc39da89ba130aef698bfa/src/coreclr/jit/lsraxarch.cpp#L2422-L2425

and it would be nice if we could make this something like:
```cpp
                srcCount += BuildOperandUses(op1);
                srcCount += BuildDelayFreeUses(op2, op1);
                srcCount += BuildDelayFreeUses(op3, op1);
                srcCount += BuildDelayFreeUses(op4, op1);
```

Rather than needing to do something like the following:
```cpp
                srcCount += BuildOperandUses(op1);
                srcCount += GenTree::NodesAreEquivalentLeaves(op2, op1) ? BuildOperandUses(op2) : BuildDelayFreeUses(op2, op1);
                srcCount += GenTree::NodesAreEquivalentLeaves(op3, op1) ? BuildOperandUses(op3) :BuildDelayFreeUses(op3, op1);
                srcCount += GenTree::NodesAreEquivalentLeaves(op4, op1) ? BuildOperandUses(op4) :BuildDelayFreeUses(op4, op1);
```

Reply:
@tannergooding - once again thanks for the insights. The change was easier than I thought because `BuildDelayFreeUses()` captures those scenarios.